### PR TITLE
Improve `fromString` to read time series with missing values

### LINKED ISSUE #118: Csv.toCsv - work with Seq[TimeSeries[A]]
-maybe create `toComplexCsv[A](xs:Seq[TimeSeries[A]] ) : String` next to `toCsv[A](xs: TimeSeries[A]): String`

### PR DESCRIPTION
**Pull request by alexander-myltsev**
Fixes https://github.com/carldata/timeseries/issues/118

### CODE REVIEW

**Comment by alexander-myltsev** [comment]
@wenta please, start from that PR

File: shared/src/main/scala/carldata/series/Csv.scala [hunk 1]
```diff
@@ -33,22 +38,61 @@ object Csv {
...
-    val data = str.split("\n").tail.map { line =>
-      val tokens = line.split(",")
-      (LocalDateTime.parse(tokens(0), dateFormatter).toInstant(ZoneOffset.UTC), tokens.tail.map(_.toDouble).toVector)
-    }.toVector
-
-    data.unzip._2
-      .transpose
-      .map(vs => TimeSeries(data.unzip._1, vs))
+    val lines = str.split(newLineDelimeter)
+    require(lines.nonEmpty)
```
Comment (wenta) [inline shared/src/main/scala/carldata/series/Csv.scala:42]: I think it is better to return empty sequence if lines are empty.

File: shared/src/main/scala/carldata/series/Csv.scala [hunk 2]
```diff
@@ -33,22 +38,61 @@ object Csv {
...
-    val lines = xs.dataPoints.map(x => LocalDateTime.ofInstant(x._1, ZoneOffset.UTC)
-      .format(DateTimeFormatter.ISO_LOCAL_DATE_TIME) + "," + x._2.toString)
-    header + "\n" + lines.mkString("\n")
+  def toCsv[A](xs: TimeSeries[A])(implicit num: Numeric[A]): String = {
+    toComplexCsv(Seq(xs))
+  }
+
+  /** Write Sequence of TimeSeries to CSV string */
+  def toComplexCsv[A](tss: Seq[TimeSeries[A]])(implicit num: Numeric[A]): String = {
+    require(tss.nonEmpty)
```
Comment (wenta) [inline shared/src/main/scala/carldata/series/Csv.scala:76]: Why throwing exception on empty Seq? I think it is bad for production code. Just return empty string if tss.isEmpty.

File: shared/src/main/scala/carldata/series/Csv.scala [hunk 3]
```diff
@@ -33,22 +38,61 @@ object Csv {
...
-      val tokens = line.split(",")
-      (LocalDateTime.parse(tokens(0), dateFormatter).toInstant(ZoneOffset.UTC), tokens.tail.map(_.toDouble).toVector)
-    }.toVector
-
-    data.unzip._2
-      .transpose
-      .map(vs => TimeSeries(data.unzip._1, vs))
+    val lines = str.split(newLineDelimeter)
+    require(lines.nonEmpty)
+    val header = lines.head
```
Comment (wenta) [inline shared/src/main/scala/carldata/series/Csv.scala:43]: `header` is used only once, there is no need to create new value

**Comment by wenta** [comment]
Tests failing on my machine, I will finish review when I figure out what is going on.

**Comment by alexander-myltsev** [comment]
@wenta , `sbt clean test` maybe? :) Please, also fix Travis config if needed.

**Comment by wenta** [comment]
`sbt clean test` didn't help. btw only this new tests for Csv failed

File: shared/src/main/scala/carldata/series/Csv.scala [hunk 4]
```diff
@@ -33,22 +38,61 @@ object Csv {
...
+    val columnsCount = header.split(tokensDelimeter) match {
```
Comment (wenta) [inline shared/src/main/scala/carldata/series/Csv.scala:44]: This is too complicated. Please use just `val columnsCount = header.split(tokensDelimeter).length` for now.

**Comment by alexander-myltsev** [comment]
What's the error message?

**Comment by wenta** [comment]
it fails because of this: ```throw new IllegalArgumentException("Wrong input string format"```
for example:
``` - should read single series from string *** FAILED ***
[info]   java.lang.IllegalArgumentException: Wrong input string format
[info]   at carldata.series.Csv$.fromString(Csv.scala:50)
[info]   at carldata.series.CsvTest.$anonfun$new$1(CsvTest.scala:17)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)```

File: shared/src/main/scala/carldata/series/Csv.scala [hunk 5]
```diff
@@ -33,22 +38,61 @@ object Csv {
...
+      case Array("time", "value") => 2
+      case arr if arr.size > 2 && arr(0) == "time" &&
+        arr.zipWithIndex.tail.forall { case (p, i) => p == s"value$i" } => arr.size
+      case _ => throw new IllegalArgumentException("Wrong input string format")
+    }
+    val valueColumnsCount = columnsCount - 1
+
+    val data = Vector.fill(valueColumnsCount)(ArrayBuffer.empty[(Long, Double)])
```
Comment (wenta) [inline shared/src/main/scala/carldata/series/Csv.scala:52]: I think we do not have to use mutable ArrayBuffer if `data` is result of recursive function.

**Review by wenta — CHANGES_REQUESTED** [review]

**Pull request closed without merging**